### ISSUE TITLE
Proposed more computationally-efficient method of computing the logit z in the tensorflow.keras module.

### ISSUE THREAD

**Issue by sjlearmonth** [tensorflow/tensorflow#106085]
### Issue type

Performance

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.20.0

### Custom code

Yes

### OS platform and distribution

macOS Tahoe 26.1 NumPy 2.3.5

### Mobile device

Not Applicable

### Python version

3.11.6

### Bazel version

Not Applicable

### GCC/compiler version

Python interpreter 3.11.6

### CUDA/cuDNN version

Not Applicable

### GPU model and memory

Not Applicable

### Current behavior?

I would like to propose a modification to how the logit z is computed in the tensorflow.keras module.

Traditionally, s is computed from;

`z = w * x + b`

However, by re-writing this as;

`z = w * x + b * 1`

we get the same result but slightly faster when computing z from the dot product;

`z = numpy.dot(wb, x1s)`

where `wb` is the 2D numpy array `w` with a column of parameter bias `b` values appended to the right-end of `w` for each unit in a layer. And `x1s` is the 2D numpy array `x` with an additonal row of ones, hence the `1s` or ones, appended to the bottom of `x`.

This means the logit `z` can be computed by calling;

`z = numpy.dot(wb, x1s)`

There is an approximate decrease in the computation time of 20-25% using this implementation when computing the dot product with numpy 2D arrays of shape (1000, 1000) having float values as array elements.

I have included my Python3 code for you to test and explore this efficiency saving in computation time which will also make forward prop compuation faster in training deep learning neural networks.

### Standalone code to reproduce the issue

```shell
import numpy as np
import timeit
import tensorflow as tf
import sys

print(f"Numpy version: {np.__version__}")
print(f"TensorFlow version: {tf.__version__}")
print(f"Python version: {sys.version}\n")
print("Author: Stephen J Learmonth.")
print("Date: 11th December 2025.\n")

rows = cols = 1000

trad_w = (np.random.random(size=(rows, cols)) - 0.5) * 20
trad_x = (np.random.random(size=(rows, cols)) - 0.5) * 20
trad_b = (np.random.random(size=(rows, 1)) - 0.5) * 20

elapsed_times_trad = timeit.repeat(stmt="trad_z = np.dot(trad_w, trad_x) + trad_b",
                                   setup="import numpy as np",
                                   repeat=6,
                                   number=1000,
                                   globals=globals())

min_elapsed_time_trad = min(elapsed_times_trad)

print("----------- Traditional Computation of z = w * x + b --------\n")
print("                    w.shape = (1000, 1000)                   ")
print("                    x.shape = (1000, 1000)                   ")
print("                    b.shape = (1000, 1)                      \n")

print(f"Traditional computation of z: Min elapsed execution time: {np.round((min_elapsed_time_trad / 1000)*1e3, 2)} msecs.\n")

alt_wb = np.append(trad_w, (np.random.random(size=(rows, 1)) - 0.5) * 20, axis=1)
alt_x1s = np.append(trad_x, np.ones((1, cols)), axis=0)

elapsed_times_alt = timeit.repeat(stmt="alt_z = np.dot(alt_wb, alt_x1s)",
                                  setup="import numpy as np",
                                  repeat=6,
                                  number=1000,
                                  globals=globals())

min_elapsed_time_alt = min(elapsed_times_alt)

performance_increase_pcnt = ((min_elapsed_time_trad - min_elapsed_time_alt) / min_elapsed_time_trad) * 100

print("---------------- Alternate Computation of z = wb * x1s --------------\n")
print("                    wb.shape = (1000, 1001)                        ")
print("                    x1s.shape = (1001, 1000)                       \n")

print(f"Alternate computation of z: Min elapsed execution time: {np.round((min_elapsed_time_alt / 1000)*1e3, 2)} msecs.\n")
print(f"Performance increase: {performance_increase_pcnt:.2f}%.\n")
```

### Relevant log output

```shell
Numpy version: 2.3.5
TensorFlow version: 2.20.0
Python version: 3.11.6 (v3.11.6:8b6ee5ba3b, Oct  2 2023, 11:18:21) [Clang 13.0.0 (clang-1300.0.29.30)]

Author: Stephen J Learmonth.
Date: 11th December 2025.

----------- Traditional Computation of z = w * x + b --------

                    w.shape = (1000, 1000)                   
                    x.shape = (1000, 1000)                   
                    b.shape = (1000, 1)                      

Traditional computation of z: Min elapsed execution time: 3.05 msecs.

---------------- Alternate Computation of z = wb * x1s --------------

                    wb.shape = (1000, 1001)                        
                    x1s.shape = (1001, 1000)                       

Alternate computation of z: Min elapsed execution time: 2.42 msecs.

Performance increase: 20.58%.


Process finished with exit code 0
```

**Comment by sjlearmonth** [comment]
Update code to confirm full equality between the 2D NumPy arrays whose elements are the `z` logits computed in both the traditional way and in the new, alternatively and faster way.

**Comment by sjlearmonth** [comment]
I'm closing this issue to make this suggestion as a contributor instead.
